src/include/stir/IO/InputFileFormatSAFIR.h

```diff
@@ -0,0 +1,147 @@
+/* InputFileFormatSAFIR.h
+
+ Class defining input file format for coincidence listmode data for SAFIR. 
+ Jannis Fischer
+ jannis.fischer@cern.ch
+
+	Copyright 2015 ETH Zurich, Institute of Particle Physics
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+ */
+
+/*!
+
+  \file
+  \ingroup listmode
+  \brief Declaration of class stir::SAFIRCListmodeInputFileFormat
+
+  \author Jannis Fischer
+*/
+
+#ifndef __local_SAFIR_InputFileFormatSAFIR_H__
+#define __local_SAFIR_InputFileFormatSAFIR_H__
+
+#include <cstring>
+#include <string>
+#include <iostream>
+
+#include "stir/IO/InputFileFormat.h"
+#include "stir/utilities.h"
+#include "stir/ParsingObject.h"
+
+#include "stir/listmode/CListRecordSAFIR.h"
+#include "stir/listmode/CListModeDataSAFIR.h"
+
+START_NAMESPACE_STIR
+
+//! Class for reading SAFIR coincidence listmode data.
+class SAFIRCListmodeInputFileFormat : public InputFileFormat<CListModeData>, public ParsingObject
+{
+public:
+	SAFIRCListmodeInputFileFormat() : did_parsing(false) {}
+	virtual const std::string get_name() const
+	{
+		return "SAFIR Coincidence Listmode File Format";
+	}
+	
+	//! Checks in binary data file for correct signature.
+	virtual bool can_read(const FileSignature& signature, std::istream& input ) const
+	{
+		return this->actual_can_read(signature, input);
+	}
+
+	//! Checks in binary data file for correct signature (can be either "SAFIR CListModeData" or "MUPET CListModeData").
+	virtual bool can_read( const FileSignature& signature, const std::string& filename) const
+	{
+		actual_do_parsing(filename);
```

Comment:
I think you are missing something :-;

actual_do_parsing needs to return Succeeded or so to if the filename it passed was actually "parsable" (i.e. a SAFIR header file).

To understand this, supposed I pass a filename of some binary list-mode fileformat (or really any non-SAFIR header) to `InputFileFormatSAFIR::can_read`.  it will call `actual_do_parsing` which will go horribly wrong, but won't tell `can_read`, which would therefore just proceed and try to use the crystal map filename anyway.

Because of the same reason, I think it's safe to check the FileSignature first before actually attempting to do any parsing at all.